kbmd/zpool_create.c

```diff
@@ -0,0 +1,334 @@
+{
+	errf_t *ret = ERRF_OK;
+	const uint8_t *sys_piv_guid = NULL;
+	char gstr[GUID_STR_LEN] = { 0 };
+
+	if (guid != NULL) {
+		guidtohex(guid, gstr, sizeof (gstr));
+	} else {
+		(void) strlcpy(gstr, "(not set)", sizeof (gstr));
+	}
+
+	(void) bunyan_trace(tlog, "try_sys_piv: enter",
+	    BUNYAN_T_STRING, "guid", gstr,
+	    BUNYAN_T_END);
+
+	ASSERT(MUTEX_HELD(&piv_lock));
+
+	if (sys_piv == NULL || guid == NULL) {
+		(void) bunyan_trace(tlog,
+		    "sys piv not set or guid not specified",
+		    BUNYAN_T_END);
+		return (B_FALSE);
+	}
+
+	sys_piv_guid = piv_token_guid(sys_piv->kt_piv);
+
+	if (bcmp(sys_piv_guid, guid, GUID_LEN) != 0) {
+		(void) bunyan_trace(tlog, "specified guid is not sys piv",
+		    BUNYAN_T_STRING, "guid", gstr,
+		    BUNYAN_T_END);
+		return (B_FALSE);
+	}
+
+	if (sys_piv->kt_rtoken.rt_val == NULL && rtoken == NULL) {
+		(void) bunyan_trace(tlog, "System PIV not set",
+		    BUNYAN_T_END);
+		return (B_FALSE);
+	}
+
+	if (sys_piv->kt_rtoken.rt_val != NULL)
+		return (B_TRUE);
+
+	if ((ret = set_piv_rtoken(sys_piv, rtoken)) != ERRF_OK) {
+		/*
+		 * This can only fail due to no memory, so we don't
+		 * care about the exact error message.
+		 */
+		errf_free(ret);
+		return (B_FALSE);
+	}
+	return (B_TRUE);
+}
+
+/*
+ * Verify we have a token to use.  If we've already setup the token,
+ * we don't require the token to be wiped and re-initialized + setup to
+ * retry recreating the zpool.
+ */
+static errf_t *
+kbmd_assert_token(const uint8_t *guid, const recovery_token_t *rtoken,
+    kbmd_token_t **restrict ktp, struct ebox_tpl **restrict rcfgp)
+{
+	errf_t *ret = ERRF_OK;
+
+	ASSERT(MUTEX_HELD(&piv_lock));
+
+	*ktp = NULL;
+	*rcfgp = NULL;
+
+	/*
+	 * If the system PIV has been designated, and is usable, we
+	 * use that.
+	 */
+	if (try_sys_piv(guid, rtoken)) {
+		(void) bunyan_debug(tlog, "Using system token",
+		    BUNYAN_T_STRING, "piv_guid",
+		    piv_token_guid_hex(sys_piv->kt_piv),
+		    BUNYAN_T_END);
+
+		*ktp = sys_piv;
+		kbmd_token_free(incomplete_tok);
+		incomplete_tok = NULL;
+		return (ERRF_OK);
+	}
+
+	if ((ret = try_guid(guid, rtoken, ktp)) != ERRF_OK)
+		return (ret);
+
+	if (guid != NULL) {
+		ASSERT3P(*ktp, !=, NULL);
+
+		(void) bunyan_debug(tlog, "Using supplied PIV token",
+		    BUNYAN_T_STRING, "piv_guid",
+		    piv_token_guid_hex((*ktp)->kt_piv),
+		    BUNYAN_T_END);
+
+		kbmd_token_free(incomplete_tok);
+		incomplete_tok = NULL;
+		kbmd_set_token(*ktp);
+		return (ERRF_OK);
+	}
+
+	if (incomplete_tok != NULL) {
+		(void) bunyan_debug(tlog,
+		    "Using token info from previous attempt(s)",
+		    BUNYAN_T_STRING, "piv_guid",
+		    piv_token_guid_hex(incomplete_tok->kt_piv),
+		    BUNYAN_T_END);
+
+		if ((ret = register_pivtoken(incomplete_tok,
+		    rcfgp)) != ERRF_OK ) {
+
+			(void) bunyan_error(tlog,
+			    "Failed to register pivtoken, token data saved for "
+			    "retry", BUNYAN_T_END);
+
+			return (ret);
+		}
+
+		*ktp = incomplete_tok;
+		incomplete_tok = NULL;
```

Comment:
Does this ever result in a leak?  The same switcheroo is done in many places, maybe there should be a function that does this.

Reply:
Yes and no :)

The difficulty here is once a PIV token is initialized, kbmd is the only thing that knows the pin until after it successfully registers the new PIV token with KBMAPI. If the KBMAPI registration fails for any reason, we want to keep the info for the new PIV token around, and have subsequent `sdc-server setup ... encryption_enabled=true` attempts to be able to retry without having to factory reset the PIV token first. It's not perfect -- if you hit such a situation and kbmd is restarted, you have to do the PIV reset, but that seems likely to be a much smaller subset of potential issues than something where you'd just re-run `sdc-server setup`.  We do support some additional arguments to `kbmadm create-pool`, but they're never used by `sdc-server setup` and are intended for troubleshooting (hopefully never necessary, but it seemed reasonable to make sure we leave options there just in case).

Reply:
Once the PIV token is initialized, it's set as the system token (since filesystem/smartdc won't do it during the initial setup).

Reply:
confirmed